[FIX] website: prevent cow if editing generic page with diverged URL

rdeodoo opened the pull request:
In the case of a generic page and its specific page: When the specific page change its URL, its view will still have the sam key as the generic page despite not being served on the same URL anymore and not really being the same ressource anymore. This is a bit weird conceptually.
Indeed, the generic page is not shadowed by the specific one anymore.

This is an issue because when going on the generic page and editing it, the save will actually "bug": it will not write on the generic view that you edited but the save will instead be "redirected" (through the COW mechanism) to the specific view.
The editor will look like it did not save your change, while in fact it actually erased, behind the scene, the specific page content and replaced it by the one you tried to save on the generic page. You really just lost your specific page content without knowing it.

This is a corner case of the COW mechanism and the holy grail rule of multi website (editing a website should only impact itself and not other websites): we serve both the generic content and its specific content on the website, which should never be possible except for this particular case of a view being linked to a page and the specific view's page having its URL changed.

Note that an apparently good solution would be to adapt the view key to reflect its new page URL (as when you create a new page), but it's not possible. We don't do that for a simple reason: the page view's key could be xpath'd. And it is even the case for the contactus page. Even if it is not a very legit flow to xpath a page view, because it makes (most of the time) that page not possible to edit, we have to support it.

Also note that a generic page is not something supposed to exist, even if we support it (and use it for /contactus), it's best to avoid it:
1. You can't do that through the UI unless going into debug mode
2. It does not make a lot of sense to share a whole page content on multiple websites, as it would duplicated content (bad for SEO).

Steps to reproduce:
- Go to /contactus and enter edit mode
- Change something like add "WEBSITE1" somewhere in the page
- Open the page properties and change the url from /contactus to /contactuswebsite1
- Go to /contactus, which will still be available and now show the generic page
- You will not see the "WEBSITE1" you added, which is what is expected as this is the generic page, the "WEBSITE1" is on the /contactuswebsite1 page
- Enter edit mode and add something, like "GENERIC", save
- The page will reload (as you saved) but your "GENERIC" will disappear, looking like it did not save your change
- Now go to /contactuswebsite1, you will see that the "WEBSITE1" word is gone, and the "GENERIC" word is there

On top of the following OPWs, the bug was also reported internally on discord.
opw-3760257
opw-3473923
task-3476840

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/159297).

rdeodoo commented:
@qsm-odoo The diff is simpler than what I initially planed, so I switched to Odoo 15.
Functionally it seems to be the best compromise, I don't see many possibilities. Conveniently, it's also the easiest approach technically.

qsm-odoo commented on addons/website/models/ir_ui_view.py:
> Note that an apparently good solution would be to adapt the view key to
reflect its new page URL (as when you create a new page), but it's not
possible. We don't do that for a simple reason: the page view's key
could be xpath'd. And it is even the case for the contactus page.
Even if it is not a very legit flow to xpath a page view, because it
makes (most of the time) that page not possible to edit, we have to
support it.

To review, explanation's not right as discussed.

> as it would duplicated content (bad for SEO).

would **be**.

(Funny, I never thought that having the feature of making a generic page could actually be bad.)

rdeodoo replied:
```diff
Note that an apparently good solution would be to adapt the view key to
reflect its new page URL (as when you create a new page), but it's not
possible. We don't do that for a simple reason: the page view's key
- could be xpath'd. And it is even the case for the contactus page.
+ could be t-called'd. And it is even the case for the contactus page.
- Even if it is not a very legit flow to xpath a page view, because it
- makes (most of the time) that page not possible to edit, we have to
- support it.
```

qsm-odoo commented on addons/website/models/ir_ui_view.py:
```diff
@@ -522,7 +522,19 @@ def save(self, value, xpath=None):
             if website_specific_view:
-                self = website_specific_view
+                if (
+                    website_specific_view.page_ids
+                    and website_specific_view.page_ids[0].url != self.page_ids[0].url
```
Not sure why we do but we do have `first_page_id` which maybe is made for accessing the only page linked to the view? :thinking:

rdeodoo replied:
We need this field to be able to add the (clickable) relationship inside the form view of an `ir.ui.view` to it's `website.page`.
Due to the nature of our ORM, it's a o2m (inverse of the m2o), even if it does not make sense to have multiple page for a given view, the ORM allows it (and never see a point to try to write custom code to prevent it).

It's only used for this at the moment but indeed we can probably use it here

qsm-odoo commented on addons/website/models/ir_ui_view.py:
```diff
@@ -522,7 +522,19 @@ def save(self, value, xpath=None):
+                    website_specific_view.page_ids
+                    and website_specific_view.page_ids[0].url != self.page_ids[0].url
+                ):
+                    # The case here is when a generic page is edited after its
+                    # specific page has a different URL. In this case, the
+                    # generic page can still be accessed since the specific one
+                    # does not shadow it anymore. In such a case, we need the
+                    # write to be done on the edited generic page and not target
+                    # the specific one.
+                    self = self.with_context(no_cow=True)
```
Not sure I like this, IMO once you change the URL of the generic page it should not be available for your current website anymore (the routing should see that you made a specific page with a different URL). So this case here should not exist. IMO the flow may look more buggy than before:
- Go to website 1 /contactus page
- Enter edit mode, change the URL used as email_to for the form, save
- Open page properties, change the URL
- Go to /contactus again, you land on the generic page... but you have no way of knowing it, it's just a different email. Enter edit mode, add a banner at the top... now you edited all your ten websites but not the one you actually wanted clicking on "Contact us" from the page manager, not even looking at the URL :shrug: 

I guess this code could exist as a fallback for a case which is not supposed to exist... but the main problem IMO is being able to land on the generic page whose URL was changed. And I guess that if we do want to implement that, we cannot indeed change the keys (even in master). In fact why is it not already working if we base the "is there a specific view?" on keys?

Who knows, maybe in master, we should not allow the generic page feature entirely. As you said, probably bad as SEO always :shrug: So we could have something like "when creating a website, a page is created for contact us and homepage but they are specific without generic automatically" :shrug: Although, I am sure the PS tech could have used that feature, bad for SEO or not? :shrug:

rdeodoo replied:
> Although, I am sure the PS tech could have used that feature, bad for SEO or not? 🤷

Yup, it's probably something we need to keep, it's probably very practical in some cases where you don't care about SEO or don't even index the page.

> IMO the flow may look more buggy than before:

I don't think it's worst than keep trying to save changes that are actually looking like they are never saved and you keep losing your changes (while it is just writing on another page behind the scene), which is what the 2-3 people complained when they reported the issue, but yeah, it's probably in both cases at the very high of the "terrible UX" scale :grimacing: 

Even if I think you mostly do visual diff (as explained in my step to reproduce and reported by people) than hidden diff (your case about email option), so it really looks "wtf is going on".

I'll try to figure another way, to give the editor a warning upfront like a banner or something on the generic page when we detect a specific counterpart.
Or maybe try to alter the routing to not let you access the page at all once it's "COW'd" and url changed.
Even if it seems very weak and fragile to consider such pages as "Generic and specific to each other" just because they got the same key..
For the user, they changed their url and their content, there is nothing left in common between the pages.

It's why I made the fix that way, for me it seems quite correct to say "you have modified the page URL, and most of the time it's content too, those are different pages now".

So about what you say here:

> Go to /contactus again, you land on the generic page... but **you have no way of knowing it**, it's just a different email.

Imo there is nothing to know, you made those pages independent and not related anymore. You most of the time will see a visual diff and should notice it.

But I got your opinion.. Not an easy one :grimacing: 
I think the banner is the way to go: you can access the page, but the system is letting you know that this is linked to another specific page. We will need to find a good wording.

Or even simpler, when entering edit mode on such a generic page shadowed by a specific page (again, specific for odoo technical code, but not for lambda users, those have nothing in common anymore), we show an alert to the user (could be blocking).

Let's discuss this tomorrow IRL once you've read this to see what we decide?

rdeodoo commented:
@qsm-odoo I've made a fix to prevent the generic page access instead, it's working well from the few tests I made, I'll triple check and wait for runbot (I also pushed a test to cover it).

But I am not sure about introducing it in Odoo 15: "what if".. ?
People could have ressources shown / used on generic page where that generic page has a specific page with different URL also used.
I'd target this on Odoo 17 IMO, what do you think?

(I kept the initial fix in the meantime.)

qsm-odoo commented:
@rdeodoo I guess 17.0 is fine, although in that case, maybe restore your previous diff for 15.0? Could be in another PR :shrug:

rdeodoo commented:
Let's do that :+1: 
Done, here is the commit to be use for the 17 forward port:

<details>

<summary> Diff patch </summary>

```
From 8a2b8bbaf33ad30e5f1faff6b9c76f79a341f576 Mon Sep 17 00:00:00 2001
From: Romain Derie <rde@odoo.com>
Date: Mon, 22 Apr 2024 09:16:09 +0200
Subject: [PATCH] [FIX] website: prevent generic page access if cow + cow url
 changed

In the case of a generic page and its specific page:
When the specific page change its URL, its view will still have the same
key as the generic page despite not being served on the same URL anymore
and not really being the same ressource anymore. This is a bit weird
conceptually.
Indeed, the generic page is not shadowed by the specific one anymore.

This is an issue because when going on the generic page and editing it,
the save will actually "bug": it will not write on the generic view that
you edited but the save will instead be "redirected" (through the COW
mechanism) to the specific view.
The editor will look like it did not save your change, while in fact it
actually erased, behind the scene, the specific page content and
replaced it by the one you tried to save on the generic page.
You really just lost your specific page content without knowing it.

This is a corner case of the COW mechanism and the holy grail rule of
multi website (editing a website should only impact itself and not other
websites): we serve both the generic content and its specific content on
the website, which should never be possible except for this particular
case of a view being linked to a page and the specific view's page
having its URL changed.

Note that an apparently good solution would be to adapt the view key to
reflect its new page URL (as when you create a new page), but it's not
possible. We don't do that for a simple reason: the page view's key
could be t-called'd.

Also note that a generic page is not something supposed to exist, even
if we support it (and use it for /contactus), it's best to avoid it:
1. You can't do that through the UI unless going into debug mode
2. It does not make a lot of sense to share a whole page content on
   multiple websites, as it would be duplicated content (bad for SEO).

Steps to reproduce:
- Go to /contactus and enter edit mode
- Change something like add "WEBSITE1" somewhere in the page
- Open the page properties and change the url from /contactus to
  /contactuswebsite1
- Go to /contactus, which will still be available and now show the
  generic page
- You will not see the "WEBSITE1" you added, which is what is expected
  as this is the generic page, the "WEBSITE1" is on the
  /contactuswebsite1 page
- Enter edit mode and add something, like "GENERIC", save
- The page will reload (as you saved) but your "GENERIC" will disappear,
  looking like it did not save your change
- Now go to /contactuswebsite1, you will see that the "WEBSITE1" word
  is gone, and the "GENERIC" word is there

This commit simply prevent the access of a generic page if it has a
specific page with a different URL.
Note:
- If the specific page has the same URL, there is nothing to do, our
  COW/multi website mechanism will just make it work, the specific will
  shadow the generic.
- The fix done here has no perf impact when we are not in that very rare
  scenario.

On top of the following OPWs, the bug was also reported internally on
discord.
[opw-3760257](https://www.odoo.com/web#id=3760257&view_type=form&model=project.task&menu_id&cids=1)
opw-3473923
task-3476840
---
 addons/website/models/ir_http.py  | 12 +++++++++++-
 addons/website/tests/test_page.py | 27 +++++++++++++++++++++++++++
 2 files changed, 38 insertions(+), 1 deletion(-)

diff --git a/addons/website/models/ir_http.py b/addons/website/models/ir_http.py
index e308d4189ddfa..a2e9a4ac07c03 100644
--- a/addons/website/models/ir_http.py
+++ b/addons/website/models/ir_http.py
@@ -296,7 +296,17 @@ def _search_page(comparator='='):
             page.browse([page.id] + menu_pages_ids).mapped('view_id.name')
             request.website.menu_id
 
-        if page and (request.website.is_publisher() or page.is_visible):
+        if (
+            page
+            and (request.website.is_publisher() or page.is_visible)
+            and (
+                # If a generic page (niche case) has been COWed and that COWed
+                # page received a URL change, it should not let you access the
+                # generic page anymore, despite having a different URL.
+                page.website_id
+                or not page.view_id._get_specific_views().filtered(lambda view: view.website_id == request.website)
+            )
+        ):
             need_to_cache = False
             cache_key = page._get_cache_key(request)
             if (
diff --git a/addons/website/tests/test_page.py b/addons/website/tests/test_page.py
index 1b200698cd12d..368c5ffab6ec1 100644
--- a/addons/website/tests/test_page.py
+++ b/addons/website/tests/test_page.py
@@ -293,6 +293,33 @@ def test_page_url_case_insensitive_match(self):
         self.assertEqual(r2.status_code, 303, "URL exists only in different casing, should redirect to it")
         self.assertTrue(r2.headers.get('Location').endswith('/page_1'), "Should redirect /Page_1 to /page_1")
 
+    def test_page_generic_diverged_url(self):
+        """ When a generic page is COW and the new COW has its url changed, the
+        generic should not be reachable anymore even if the COW page has a
+        different URL. Note that they will both still share the same key.
+        """
+        Page = self.env['website.page']
+        specific_arch = '<div>website 1 content</div>'
+        generic_page = self.page
+        generic_page.arch = '<div>content</div>'
+
+        specific_page = Page.search([('url', '=', self.page.url), ('website_id', '=', 1)])
+        self.assertFalse(specific_page, "For this test, the specific page should not exist yet")
+
+        # COW a generic page
+        generic_page.view_id.with_context(website_id=1).save(specific_arch, xpath='/div')
+        specific_page = Page.search([('url', '=', self.page.url), ('website_id', '=', 1)])
+        self.assertEqual(specific_page.arch.replace('\n', ''), specific_arch)
+        self.assertEqual(generic_page.arch, '<div>content</div>')
+        # Change the URL of the specific page
+        specific_page.url = '/page_1_specific'
+        # Check that the generic page is not reachable anymore
+        r = self.url_open(specific_page.url)
+        self.assertEqual(r.status_code, 200, "Specific should be reachable")
+        r = self.url_open(generic_page.url)
+        self.assertEqual(r.status_code, 404, "Generic should not be reachable")
+
+
 @tagged('-at_install', 'post_install')
 class TestNewPage(common.TransactionCase):
     def test_new_page_used_key(self):
```

<details>

Summary:
- In Odoo 15 and Odoo 16, people will be editing what they see, so if they are on the generic page, they will edit the generic page, instead of having their write looking like it just did nothing
- In Odoo 17, generic page won't be reachable, we will prevent its access

robodoo commented:
@qsm-odoo you may want to rebuild or fix this PR as it has failed CI.

qsm-odoo commented:
@robodoo r+
